### PR TITLE
[C-2761] Add mutations to audius-query

### PR DESCRIPTION
**Pull request by dylanjeffers**
### Description

Adds "mutation" endpoint by specifying `type: 'mutation'` in endpoint options object. I dont love that mutations are tucked away with a small options update, but that was the closest i could get compared the the rtk approach of using a "builder" to specify query vs mutation. (typechecks well)
Adds optimistic updates with an api utility `updateQueryData`. This pattern is copied from rtk-query. (typechecks well)
Adds `redux-thunk` to our store enhancers, since the `updateQueryData` utility requires it.

### Dragons

The code could be much less wet. Essentially there is a lot of code that can be reused between queries and mutations, but since the signatures of the resulting hooks were so different, it was hard to find an easy way to share. Long term we can likely do this, though.
### How Has This Been Tested?

Tested with an upcoming feature to developer-apps

### CODE REVIEW

**Comment by audius-infra** [comment]
It looks like there may be some changes to native mobile code, which requires triggering a full app release. Please follow the instructions here: https://www.notion.so/audiusproject/When-to-bump-app-version-2644a8f772364a4d91f44abcba44ce0b?pvs=4. cc @nicoback2 @sliptype

**Comment by audius-infra** [comment]
It looks like there may be some changes to native mobile code, which requires triggering a full app release. Please follow the instructions here: https://www.notion.so/audiusproject/When-to-bump-app-version-2644a8f772364a4d91f44abcba44ce0b?pvs=4. cc @nicoback2 @sliptype

**Comment by audius-infra** [comment]
It looks like there may be some changes to native mobile code, which requires triggering a full app release. Please follow the instructions here: https://www.notion.so/audiusproject/When-to-bump-app-version-2644a8f772364a4d91f44abcba44ce0b?pvs=4. cc @nicoback2 @sliptype

File: packages/web/package-lock.json [hunk 1]
```diff
@@ -2578,88 +2578,6 @@
       "resolved": "https://registry.npmjs.org/@ctrl/tinycolor/-/tinycolor-3.4.1.tgz",
       "integrity": "sha512-ej5oVy6lykXsvieQtqZxCOaLT+xD4+QNarq78cIYISHmZXshCvROLudpQN3lfL8G0NL7plMSSK+zlyvCaIJ4Iw=="
     },
-    "@cykura/sdk-core": {
```
Comment (amendelsohn) [inline packages/web/package-lock.json:2581]: feels like too many lock changes

File: packages/common/src/audius-query/README.md [hunk 2]
```diff
@@ -128,6 +135,23 @@
     )
     ```
 
+1.  In your component, if your hook was generated by a "mutation" endpoint
```
[the remark this answers is not in the document]
Comment (dylanjeffers) [reply packages/common/src/audius-query/README.md:138]: yeah i wasnt sure about the best way to add the "writes" part, like the docs in react-query have the "query" and "mutation" sections clearly defined

File: packages/common/src/audius-query/createApi.ts [hunk 3]
```diff
@@ -116,6 +133,96 @@ const addEndpointToSlice = <NormalizedData>(
   }
 }
 
+const useQueryState = <Args, Data>(
```
[the remark this answers is not in the document]
Comment (dylanjeffers) [reply packages/common/src/audius-query/createApi.ts:136]: i did hook selector originally, but then went with hook to share even more, like the isEqual check, and avoiding passing state as another arg to the selector

File: packages/common/src/audius-query/README.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (amendelsohn) [reply packages/common/src/audius-query/README.md:138]: mhmm I think we should separate them

File: packages/common/src/audius-query/createApi.ts [hunk 4]
```diff
@@ -259,24 +294,35 @@ const buildEndpointHooks = <
...
           if (!apiData) {
             throw new Error('Remote data not found')
           }
 
-          const { entities, result } = normalize(
-            { [endpoint.options.schemaKey]: apiData },
-            apiResponseSchema
-          )
-          dispatch(addEntries(Object.keys(entities), entities))
+          let nonNormalizedData: any
```
[the remark this answers is not in the document]
Comment (dylanjeffers) [reply packages/common/src/audius-query/createApi.ts:305]: so the main thing here was to allow schemaKey to be optional, and then needed a way to either get non-normalized-data from the schema denorm, or just itself. what do you think should be the approach here?

File: packages/common/src/audius-query/types.ts [hunk 5]
```diff
@@ -18,12 +22,33 @@ export type Api<EndpointDefinitions extends DefaultEndpointDefinitions> = {
...
-    >}`]: (
-      fetchArgs: Parameters<EndpointDefinitions[Property]['fetch']>[0],
-      options?: QueryHookOptions
-    ) => QueryHookResults<
-      Awaited<ReturnType<EndpointDefinitions[Property]['fetch']>>
-    >
+    >}`]: EndpointDefinitions[Property]['options']['type'] extends 'mutation'
+      ? () => [
+          (
+            fetchArgs: Parameters<EndpointDefinitions[Property]['fetch']>[0],
```
[the remark this answers is not in the document]
Comment (dylanjeffers) [reply packages/common/src/audius-query/types.ts:28]: yeah the conditional here... i was like SURELY this cant work

File: packages/common/src/audius-query/types.ts [hunk 6]
```diff
@@ -38,13 +63,20 @@ type EndpointOptions = {
   idArgKey?: string
   idListArgKey?: string
   permalinkArgKey?: string
-  schemaKey: string
+  schemaKey?: string
```
[the remark this answers is not in the document]
Comment (dylanjeffers) [reply packages/common/src/audius-query/types.ts:66]: yeah and i see a case where there are no schema objects in a request, like dev-apps?

File: packages/web/package-lock.json (hunk 1 above)
Comment (dylanjeffers) [reply packages/web/package-lock.json:2581]: yeah agreed. i rverted to main, removed node_modules, ran i, and this is what happenend. i think it's cause there were recently a lot of probably unwanted lock changes on a recent commit?

File: packages/common/src/audius-query/createApi.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (amendelsohn) [reply packages/common/src/audius-query/createApi.ts:305]: I mean I think the structure is fine, we can just use a different name to make it clear it's separate from the nonNormalizedData from line 250.

Not a big deal though, so use whichever you feel is more readable

[the remark this answers is not in the document]
Comment (dylanjeffers) [reply packages/common/src/audius-query/createApi.ts:305]: ahh right right, sure thing

File: packages/web/package-lock.json (hunk 1 above)
Comment (amendelsohn) [reply packages/web/package-lock.json:2581]: ok yeah that makes sense. Might as well go back to goo

File: packages/common/src/audius-query/README.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (dylanjeffers) [reply packages/common/src/audius-query/README.md:138]: yeah lets do it

**Comment by audius-infra** [comment]
It looks like there may be some changes to native mobile code, which requires triggering a full app release. Please follow the instructions here: https://www.notion.so/audiusproject/When-to-bump-app-version-2644a8f772364a4d91f44abcba44ce0b?pvs=4. cc @nicoback2 @sliptype

**Pull request merged**